File: agent/src/main/java/com/teamscale/jacoco/agent/DelayedLogger.java
package com.teamscale.jacoco.agent;

import com.teamscale.report.util.ILogger;
import org.slf4j.Logger;

import java.util.ArrayList;
import java.util.List;

/**
 * A logger that buffers logs in memory and writes them to the actual logger at a later point.
 * This is needed when stuff needs to be logged before the actual logging framework is initialized.
 */
public class DelayedLogger implements ILogger {

	/** List of log actions that will be executed once the logger is initialized. */
	private final List<ILoggerAction> logActions = new ArrayList<>();

	@Override
	public void debug(String message) {
		logActions.add(logger -> logger.debug(message));
	}

	@Override
	public void info(String message) {
		logActions.add(logger -> logger.info(message));
	}

	@Override
	public void warn(String message) {
		logActions.add(logger -> logger.warn(message));
	}

	@Override
	public void warn(String message, Throwable throwable) {
		logActions.add(logger -> logger.warn(message, throwable));
	}

	@Override
	public void error(Throwable throwable) {
		logActions.add(logger -> logger.error(throwable.getMessage(), throwable));
	}

	@Override
	public void error(String message, Throwable throwable) {
		logActions.add(logger -> logger.error(message, throwable));
	}

	/** Writes the logs to the given slf4j logger. */
	public void logTo(Logger logger) {
		logActions.forEach(action -> action.log(logger));
	}

	/** An action to be executed on a logger. */
	private interface ILoggerAction {

		/** Executes the action on the given logger. */
		void log(Logger logger);

	}
}

File: agent/src/main/kotlin/com/teamscale/jacoco/agent/DelayedLogger.kt
package com.teamscale.jacoco.agent

import com.teamscale.report.util.ILogger
import org.slf4j.Logger
import java.util.function.Consumer

/**
 * A logger that buffers logs in memory and writes them to the actual logger at a later point. This is needed when stuff
 * needs to be logged before the actual logging framework is initialized.
 */
class DelayedLogger : ILogger {
	/** List of log actions that will be executed once the logger is initialized.  */
	private val logActions = mutableListOf<(Logger) -> Unit>()

	override fun debug(message: String) {
		logActions.add { it.debug(message) }
	}

	override fun info(message: String) {
		logActions.add { it.info(message) }
	}

	override fun warn(message: String) {
		logActions.add { it.warn(message) }
	}

	override fun warn(message: String, throwable: Throwable?) {
		logActions.add { it.warn(message, throwable) }
	}

	override fun error(throwable: Throwable) {
		logActions.add { it.error(throwable.message, throwable) }
	}

	override fun error(message: String, throwable: Throwable?) {
		logActions.add { it.error(message, throwable) }
	}

	/**
	 * Logs an error and also writes the message to [System.err] to ensure the message is even logged in case
	 * setting up the logger itself fails for some reason (see TS-23151).
	 */
	fun errorAndStdErr(message: String?, throwable: Throwable?) {
		System.err.println(message)
		logActions.add { it.error(message, throwable) }
	}

	/** Writes the logs to the given slf4j logger.  */
	fun logTo(logger: Logger) {
		logActions.forEach { it(logger) }
		logActions.clear()
	}
}

